flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java

```diff
@@ -0,0 +1,449 @@
+            // this will be enqueued in the enumerator thread, so it will actually run after this
+            // method (the snapshot operation) is complete!
+            context.runInCoordinatorThread(this::fullFillPendingRequests);
+
+            return super.snapshotState();
+        }
+
+        private void fullFillPendingRequests() {
+            for (int subtask : pendingRequests) {
+                super.handleSplitRequest(subtask, null);
+            }
+            pendingRequests.clear();
+        }
+    }
+
+    private static class TestingNumberSequenceSource extends NumberSequenceSource {
+
+        private final int numSplits;
+
+        public TestingNumberSequenceSource(long from, long to, int numSplits) {
+            super(from, to);
+            this.numSplits = numSplits;
+        }
+
+        @Override
+        public SplitEnumerator<NumberSequenceSplit, Collection<NumberSequenceSplit>>
+                createEnumerator(final SplitEnumeratorContext<NumberSequenceSplit> enumContext) {
+            final List<NumberSequenceSplit> splits =
+                    splitNumberRange(getFrom(), getTo(), numSplits);
+            return new AssignAfterCheckpointEnumerator<>(enumContext, splits);
+        }
+    }
+
+    // ------------------------------------------------------------------------
+    //  Utils for RPC intercepting cluster
+    // ------------------------------------------------------------------------
+
+    private static class OperatorEventRpcHandler {
+
+        private final FilteredRpcAction actionForFilteredEvent;
+        private final Set<Integer> eventsToFilter;
+        private int eventNum;
+
+        OperatorEventRpcHandler(FilteredRpcAction actionForFilteredEvent, int... eventsToFilter) {
+            this(
+                    actionForFilteredEvent,
+                    IntStream.of(eventsToFilter).boxed().collect(Collectors.toSet()));
+        }
+
+        OperatorEventRpcHandler(
+                FilteredRpcAction actionForFilteredEvent, Set<Integer> eventsToFilter) {
+            this.actionForFilteredEvent = actionForFilteredEvent;
+            this.eventsToFilter = eventsToFilter;
+        }
+
+        CompletableFuture<Acknowledge> filterCall(
+                ExecutionAttemptID task,
+                OperatorID operator,
+                SerializedValue<OperatorEvent> evt,
+                TriFunction<
+                                ExecutionAttemptID,
+                                OperatorID,
+                                SerializedValue<OperatorEvent>,
+                                CompletableFuture<Acknowledge>>
+                        rpcHandler) {
+
+            final Object o;
+            try {
+                o = evt.deserializeValue(getClass().getClassLoader());
+            } catch (Exception e) {
+                throw new Error(e); // should never happen
+            }
+
+            if (o instanceof AddSplitEvent || o instanceof NoMoreSplitsEvent) {
```

Comment:
Do you have a suggestion for how this should look like? 
I think what is reusable is everything below the following line in the code, the parts above are specific to that special test.
```
    // ------------------------------------------------------------------------
    //  Utils for MiniCluster RPC intercepting
    // ------------------------------------------------------------------------
```

Reply:
I see that `Source Operator Event specific intercepting`. Initially, I thought it might be a good to make `AddSplitEvent` and `NoMoreSplitsEvent` explicitly in construction side of `OperatorEventRpcHandler` so readers can get more insights in first place. I am ok to either approaches given that we only intercept these two events in long time.